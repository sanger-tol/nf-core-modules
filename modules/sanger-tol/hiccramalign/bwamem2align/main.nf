process HICCRAMALIGN_BWAMEM2ALIGN {
    tag "$meta.id"
    label "process_high"

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://community-cr-prod.seqera.io/docker/registry/v2/blobs/sha256/41/41a128095da89d592889899dceed75ba0252bf9471eebee2db6ae0eb5474738a/data' :
        'community.wave.seqera.io/library/bwa-mem2_samtools:34483fa65c9842a5' }"

    input:
    tuple val(meta), path(cram), path(crai), val(chunkn), val(range), path(index), path(reference)

    output:
    tuple val(meta), path("*.bam"), emit: bam
    path "versions.yml"           , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args1 = task.ext.args1 ?: ''
    def args2 = task.ext.args2 ?: ''
    def args3 = task.ext.args3 ?: ''
    def args4 = task.ext.args4 ?: ''
    def args5 = task.ext.args5 ?: ''
    def args6 = task.ext.args6 ?: ''
    def prefix  = task.ext.prefix ?: "${cram}.${chunkn}.${meta.id}"
    // Please be aware one of the tools here required mem = 28 * reference size!!!
    """
    INDEX=`find -L ./ -name "*.amb" | sed 's/\\.amb\$//'`

    samtools cat ${args1} -r "#:${range[0]}-${range[1]}" ${cram} |\\
        samtools fastq ${args2} - |\\
        bwa-mem2 mem ${args3} -t ${task.cpus} \${INDEX} - |\\
        samtools fixmate ${args4} - - |\\
        samtools view -h ${args5} |\\
        samtools sort ${args6} -@${task.cpus} -T ${prefix}_tmp -o ${prefix}.bam -

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        samtools: \$(echo \$(samtools --version 2>&1) | sed 's/^.*samtools //; s/Using.*\$//' )
        bwamem2: \$(echo \$(bwa-mem2 version 2>&1) | sed 's/.* //')
    END_VERSIONS
    """

    stub:
    def prefix  = task.ext.prefix ?: "${cram}.${chunkn}.${meta.id}"
    """
    touch ${prefix}.bam

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        samtools: \$(echo \$(samtools --version 2>&1) | sed 's/^.*samtools //; s/Using.*\$//' )
        bwamem2: \$(echo \$(bwa-mem2 version 2>&1) | sed 's/.* //')
    END_VERSIONS
    """
}
