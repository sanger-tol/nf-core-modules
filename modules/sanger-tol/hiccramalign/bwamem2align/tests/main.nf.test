
nextflow_process {

    name "Test Process HICCRAMALIGN_BWAMEM2ALIGN"
    script "../main.nf"
    process "HICCRAMALIGN_BWAMEM2ALIGN"

    tag "modules"
    tag "modules_sangertol"
    tag "bwamem2/index"
    tag "hiccramalign"
    tag "hiccramalign/bwamem2align"

    setup {
        run("BWAMEM2_INDEX") {
            // Copied from nf-core modules on 2025-06-2025
            // Keep up-to-date when updating the module!
            script "./bwamem2.nf"
            process {
                """
                input[0] = [
                    [ id: "test" ],
                    file(params.modules_testdata_base_path + 'Meles_meles/assembly/release/mMelMel3.1_paternal_haplotype/GCA_922984935.2.subset.fasta.gz', checkIfExists: true)
                ]
                """
            }
        }
    }

    test("meles meles") {

        when {
            process {
                """
                input[0] = [
                    [ id:'test' ], // meta map
                    file(params.modules_testdata_base_path + 'Meles_meles/genomic_data/mMelMel3/hic-arima2/35528_2%231_subset.cram', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'Meles_meles/genomic_data/mMelMel3/hic-arima2/35528_2%231_subset.cram.crai', checkIfExists: true),
                    1,
                    [0, 1],
                    BWAMEM2_INDEX.out.index[1],
                    file(params.modules_testdata_base_path + 'Meles_meles/assembly/release/mMelMel3.1_paternal_haplotype/GCA_922984935.2.subset.fasta.gz', checkIfExists: true),
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("meles meles - stub") {

        options "-stub"

        when {
            process {
                """
                input[0] = [
                    [ id:'test' ], // meta map
                    file(params.modules_testdata_base_path + 'Meles_meles/genomic_data/mMelMel3/hic-arima2/35528_2%231_subset.cram', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'Meles_meles/genomic_data/mMelMel3/hic-arima2/35528_2%231_subset.cram.crai', checkIfExists: true),
                    1,
                    [0, 1],
                    BWAMEM2_INDEX.out.index[1],
                    file(params.modules_testdata_base_path + 'Meles_meles/assembly/release/mMelMel3.1_paternal_haplotype/GCA_922984935.2.subset.fasta.gz', checkIfExists: true),
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

}
