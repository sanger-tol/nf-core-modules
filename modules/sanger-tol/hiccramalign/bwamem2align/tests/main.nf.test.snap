{
    "meles meles": {
        "content": [
            "12bbc3437515580a7dcc3c0d677c225a",
            [
                "versions.yml:md5,0d682908c95be4f8e0e547e03c448d7d"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.3"
        },
        "timestamp": "2025-07-16T16:34:31.203597"
    },
    "meles meles - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "35528_2%231_subset.cram.1.test.bam:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    "versions.yml:md5,0d682908c95be4f8e0e547e03c448d7d"
                ],
                "bam": [
                    [
                        {
                            "id": "test"
                        },
                        "35528_2%231_subset.cram.1.test.bam:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,0d682908c95be4f8e0e547e03c448d7d"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.3"
        },
        "timestamp": "2025-07-16T16:31:06.217411"
    }
}